### PR TITLE
0.1.0: Initial release

### PR DESCRIPTION
**Pull request by ziggythehamster**
@johvet identified a few things prior to submitting the review that I should fix:

* [x] changelog
* [x] escape methods could be more efficient

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+/coverage/
+/doc/
+/pkg/
+/spec/reports/
+/vendor/bundle/
+/tmp/
+
+# rspec failure tracking
+.rspec_status
+Gemfile.lock
```
Comment (ziggythehamster) [inline .gitignore:15]:
```suggestion

# Gemfile.lock files should not be committed to a gem
Gemfile.lock
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
...in the Cherokee language. I chose it because I am Cherokee and thought doctor was a good translation of druid into the language. As Cherokee is polysynthetic, there are several different ways of saying doctor, but I landed on aganakti because it is pronounced very closely to how you might intuitively pronounce it in English. Like the Japanese kana, Cherokee is written and pronounced using a syllabary, rather than an alphabet. Taking the syllables ᎠᎦᎾᎦᏘ and writing them out phonetically, you have a-ga-na-ga-ti. The vowel "a" is pronounced like the a in "father" and the vowel "i" is pronounced like the ee in "seek". The consonants in this case are pronounced like you would expect in English, but g is close to the k in "skate". Now you've got something that you would in English pronounce as "ah gah nah gah tee". If you say it fast, then you end up with something close to "ah gah knock tee", and so that is how it's pronounced and usually written. [The Cherokee Nation of Oklahoma has a recording of someone slowly saying the word if you want to hear a native speaker say it](https://data.cherokee.org/Cherokee/LexiconSoundFiles/Doctor.mp3), though they've left off the optional "a-" prefix.
```
Comment (nakedsushi) [inline README.md:58]: Cool!

File: lib/aganakti/log_subscriber.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+module Aganakti
+  ##
+  # A log subscriber using the +ActiveSupport::LogSubscriber+ interface. It only logs anything if the logger
+  # has debug level enabled.
+  #
+  # @see ActiveSupport::LogSubscriber
+  # @see https://api.rubyonrails.org/classes/ActiveSupport/LogSubscriber.html ActiveSupport::LogSubscriber in the Rails docs
+  class LogSubscriber < ActiveSupport::LogSubscriber
+    ##
+    # Called by ActiveSupport notifications when there's a SQL query that executes
```
Comment (nakedsushi) [inline lib/aganakti/log_subscriber.rb:12]: Missing doc for param and return?

**Review by nakedsushi — APPROVED** [review]

File: lib/aganakti/client.rb [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    ##
+    # @return [Hash] The options that are passed to Typhoeus
+    attr_reader :typhoeus_options
+
+    ##
+    # @return [String] The URI to make queries against
+    attr_reader :uri
+
+    ##
+    # Creates a new client instance. You probably want to use {Aganakti.new} unless you need to specify your own Typhoeus options.
```
Comment (johvet) [inline lib/aganakti/client.rb:40]: We should probably mention here, that this overrides the `Accept` and `Content-Type` members within the `:headers` option.

File: lib/aganakti/log_subscriber.rb [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    def sql(event)
```
Comment (johvet) [inline lib/aganakti/log_subscriber.rb:15]: there's no YARD for the `event` argument

File: lib/aganakti/query.rb [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    # Rather than repeat the same boilerplate over and over, use a template.
+    BOOL_SETTING_METHODS.each do |setting|
+      class_eval <<-RUBY, __FILE__, __LINE__ + 1
+        def #{setting}                                                                   # def with_approximate_count_distinct
+          if executed?                                                                   #   if executed?
+            raise QueryAlreadyExecutedError,                                             #     raise QueryAlreadyExecutedError,
+                  "#{setting} cannot be set because the query has already been executed" #           "with_approximate_count_distinct cannot be set because the query has already been executed"
+          end                                                                            #   end
+                                                                                         #
+          @#{setting.sub(/\A(with|without)_/, '')} = #{setting.start_with?('with_')}     #   @approximate_count_distinct = true
```
Comment (johvet) [inline lib/aganakti/query.rb]: I think we should consider to make this regular expression a constant.

File: lib/aganakti/query/building.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      ##
+      # Builds the query parameters that's included in the query payload.
+      #
+      # @return [Array<Hash>] query parameters in Druid's expected syntax
+      # @note Druid uses SQL-format timestamps, not ISO8601 timestamps!
+      def query_parameters # rubocop:disable Metrics/CyclomaticComplexity, Metrics/MethodLength
+        @params.map do |param|
+          case param
+          when BigDecimal  then { type: 'DECIMAL', value: param.to_s('F') }
+          when DateTime    then { type: 'TIMESTAMP', value: param.to_time.utc.strftime('%F %T.%N') }
```
Comment (johvet) [inline lib/aganakti/query/building.rb]: we should consider to make the timestamp format a constant and reuse it

File: spec/lib/aganakti/client_spec.rb [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+      client = described_class.new('http://localhost', headers: { 'Accept' => 'text/xml' })
+
+      expect(client.typhoeus_options[:headers]).to include('Accept' => 'application/json')
+    end
+
+    it 'overrides Content-Type header' do
+      client = described_class.new('http://localhost', headers: { 'Content-Type' => 'text/xml' })
+
+      expect(client.typhoeus_options[:headers]).to include('Content-Type' => 'application/json')
+    end
```
Comment (johvet) [inline spec/lib/aganakti/client_spec.rb:77]: is it worth making sure any other header passed survives? eg. `headers: { 'foo' => 'bar' }` will still be there next to Accept and Content-Type.

File: spec/lib/aganakti_spec.rb [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+  end
+
+  describe '.new' do
+    context 'with an invalid URI' do
+      it 'raises a URI::InvalidURIError' do
+        expect { described_class.new('is it pronounced hydro city zone or hydrossity zone?') }.to raise_error(URI::InvalidURIError)
+      end
+    end
+
+    context 'with a FTP URI' do
```
Comment (johvet) [inline spec/lib/aganakti_spec.rb]:
```suggestion
    context 'with an FTP URI' do
```

File: spec/lib/aganakti_spec.rb [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+      let(:url) { 'https://druidserver/query' }
+
+      it 'creates a client with the expected user agent' do
+        described_class.new(url, user_agent_prefix: 'TestSuite/1')
+
+        expect(Aganakti::Client).to have_received(:new).with(anything, hash_including(headers: hash_including('User-Agent' => %r{\ATestSuite/1 Aganakti/.*})))
+      end
+    end
+
+    context 'with credentials specified in a HTTP URI but without setting the :insecure_plaintext_login option', :stubbed_client do
```
Comment (johvet) [inline spec/lib/aganakti_spec.rb]:
```suggestion
    context 'with credentials specified in an HTTP URI but without setting the :insecure_plaintext_login option', :stubbed_client do
```

File: spec/lib/aganakti_spec.rb [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+      it 'raises an Aganakti::ConfigurationError' do
+        expect { described_class.new(url) }
+          .to raise_error(Aganakti::ConfigurationError,
+                          'Credentials cannot be provided in a HTTP URI without setting the insecure_plaintext_login option. ' \
+                          'Beware that setting this option exposes your credentials to anyone on the network and should not ' \
+                          'be used outside of development.')
+      end
+    end
+
+    context 'with credentials specified in a HTTP URI and setting the :insecure_plaintext_login option', :stubbed_client do
```
Comment (johvet) [inline spec/lib/aganakti_spec.rb]:
```suggestion
    context 'with credentials specified in an HTTP URI and setting the :insecure_plaintext_login option', :stubbed_client do
```

**Review by johvet — CHANGES_REQUESTED** [review]
This is really great stuff

File: lib/aganakti/log_subscriber.rb (hunk 3 above)
Comment (ziggythehamster) [reply lib/aganakti/log_subscriber.rb:12]: It's a void, but I added the doc for `@param`

**Review by johvet — APPROVED** [review]

**Pull request merged**